feat(api): Add G-Code Parser Makefile Execution

DerekMaggio opened the pull request:
# Overview

Added support for running the G-Code Parser CLI through the `api` project Makefile. 


# Changelog

* Add device name to text mode output
* Add HTML formatting for GCodeDiffer HTML output
* Add strings_are_equal method to GCodeDiffer
* Utilize strings_are_equal method in cli.py
* Add Makefile targets to run G-Code Parser
* Add first protocol test to Makefile

# Review requests

Check if I am doing anything dumb in the Makefile

# Risk assessment

None

Commit:
Add device to SupportedTextMode output

Commit:
feat(api): G-Code Makefile Execution

* Add HTML formatting and update tests accordingly
* Add device names to testing

Commit:
feat(api): Checking diffed strings for equality

* Add strings_are_equal method to GCodeDiffer
* Utilize strings_are_equal method in cli.py to print special message if strings are the same
* Update tests

Commit:
feat(api): Add commands for parsing G-Code

Commit:
feat(api): Add first G-Code test

amitlissack commented on api/Makefile:
```diff
@@ -194,30 +194,127 @@ term:
+#		open-diff diff_file=/path/to/file.html
+.PHONY: open-diff
+open-diff:
+	google-chrome file://$(diff_file)
```
Does this have to bee `google-chrome`? Would `open` launch the default browser?

DerekMaggio replied:
I don't have open on my system :/

amitlissack reviewed (commented):
Code looks good to me. I'll test it out before +1.

amitlissack reviewed (approved):
Tested `make g-code-smoothie-protocol-run` and worked fine. 

I recommend someone who is more Makefile knowledgable approves before merging.

sotaoverride commented on .github/workflows/g-code-confirm.yaml:
```diff
@@ -0,0 +1,40 @@
+    name: 'confirm G-Code'
+    runs-on: 'ubuntu-18.04'
+    steps:
+      - uses: 'actions/checkout@v2'
```
im trying to find this action under actions for this repo? is that where its picked up from?

sotaoverride replied:
or are these some gh actions you can add straight away ??

sotaoverride replied:
figured these actions just come with github..

sotaoverride commented on .github/workflows/g-code-confirm.yaml:
```diff
@@ -0,0 +1,40 @@
+
+on:
+  # Run on any change to the api directory
+  pull_request:
```
maybe set it to trigger for all branches for now too?

sotaoverride reviewed (commented):
trying to figure if you need to add actions to this branch?

sotaoverride reviewed (commented):
do new workflows just get picked up, or you have to add them the first time around?https://github.com/Opentrons/opentrons/actions/new

sotaoverride commented on .github/workflows/g-code-confirm.yaml:
```diff
@@ -0,0 +1,40 @@
+name: G-Code-Confirm
```
missing quotes here, 'G-Code-Confirm'

sotaoverride replied:
take that back, apparently its valid to have no quotes??

sotaoverride reviewed (dismissed):
think you just had the quotes missing for you workflow name..

sotaoverride reviewed (commented):
there were a few annoying yaml syntax things - ran the gavi workflow validator. https://github.com/Opentrons/opentrons/blob/test-ghactions/.github/workflows/g-code-confirm.yaml

sotaoverride commented on .github/workflows/g-code-confirm.yaml:
```diff
@@ -0,0 +1,40 @@
+    paths:
+      -'*'
+
+  workflow_dispatch:
```
not sure how this works - did a quick read and makes me think it needs a webhook to start or something? maybe thats why this workflow isnt running when a push / pr happens ?https://docs.github.com/en/rest/reference/actions#create-a-workflow-dispatch-event

Commit:
Pull out G-Code stuff into separate Makefile


Add Github action


Trying to get the GitHub Action to trigger


Why won't this stupid action trigger


WTF mate


Removing all triggering constraints


Changing something inside api


Trying changing the action


I broke the linting


I am begging the GitHub gods to let this trigger. Please


Add noop workflow


Add Setup


Add Python Setup


YAML syntax is annoying


Indenting sucks


Add running make command


Add uploading artifact


Add support for responding with error code


Fix test


Shhh quiet cli command


Helps if I return the error message


Archive should always happen


Try sending stderr to /dev/null.


Change file names


The action can't find the file and I don't know why. Maybe it's the space


WHY WON'T THIS WORK

Laura-Danielle reviewed (approved):
lgtm
